python/cusignal/_upfirdn.py

```diff
@@ -254,58 +270,124 @@ def __call__(self,
             x_shape_a,
             h_per_phase,
             padded_len,
-            out
+            out,
         )
 
         self.stream.use()
         self.kernel(self.grid, self.block, kernel_args)
 
 
-def _get_backend_kernel(ndim, dtype, grid, block, stream, use_numba):
+def _get_backend_kernel(
+    dtype, grid, block, stream, use_numba, k_type, ndim,
+):
     if ndim > 2:
-        raise NotImplementedError(
-            "upfirdn() requires ndim <= 2")
+        raise NotImplementedError("upfirdn() requires ndim <= 2")
     elif ndim > 1 and not use_numba:
         warnings.warn(
             "CuPy backend is only implemented for ndim == 1 \
-                Running with Numba CUDA backend", UserWarning)
+                Running with Numba CUDA backend",
+            UserWarning,
+        )
         use_numba = True
 
-    if use_numba:
-        nb_stream = stream_cupy_to_numba(stream)
-        kernel = _numba_kernel_cache[(ndim, dtype.name)]
+    if not use_numba:
+        kernel = _cupy_kernel_cache[(dtype.name, k_type)]
         if kernel:
-            return kernel[grid, block, nb_stream]
+            return _cupy_upfirdn_wrapper(grid, block, stream, kernel)
+
     else:
-        kernel = _cupy_kernel_cache[(ndim, dtype.name)]
+        nb_stream = stream_cupy_to_numba(stream)
+        kernel = _numba_kernel_cache[(dtype.name, k_type)]
         if kernel:
-            return _cupy_upfirdn_wrapper(grid, block, stream, kernel)
+            return kernel[grid, block, nb_stream]
 
     raise NotImplementedError(
-        "No kernel found for ndim {}, datatype {}".format(ndim, dtype.name))
+        "No kernel found for ndim {}, datatype {}".format(ndim, dtype.name)
+    )
 
 
-def _populate_kernel_cache():
-    for numba_type, c_type in _SUPPORTED_TYPES.items():
-        # JIT compile the numba kernels, both 1d and 2d
-        sig_1d = _numba_upfirdn_signature(numba_type, 1)
-        _numba_kernel_cache[(1, str(numba_type))] = \
-            cuda.jit(sig_1d, fastmath=True)(_numba_upfirdn_1d)
+def _populate_kernel_cache(np_type, use_numba, k_type):
 
-        sig_2d = _numba_upfirdn_signature(numba_type, 2)
-        _numba_kernel_cache[(2, str(numba_type))] = \
-            cuda.jit(sig_2d, fastmath=True)(_numba_upfirdn_2d)
+    try:
+        numba_type, c_type = _SUPPORTED_TYPES[np_type]
 
+    except KeyError:
+        raise Exception("No kernel found for datatype {}".format(np_type))
+
+    if not use_numba:
+        if (str(numba_type), k_type) in _cupy_kernel_cache:  # Not work
```

Comment:
"Not work" - I presume this is a typo, and there's no work to do?

Reply:
Yes. That comment is no longer valid.